### ISSUE TITLE
[v2] Attention Masking

### ISSUE THREAD

**Issue by MikeynJerry** [Dao-AILab/flash-attention#352]
Is any plan to add attention masking support? PyTorch's version of flash attention included the ability to provide an attention mask in their [implementation](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) that is great for removing padding from the calculation.

**Comment by leizhao1234** [comment]
In fact, when you send an attention mask to PyTorch's implementation, flash attention didn't work.

**Comment by balachandarsv** [comment]
Yes, facing the same issue. @tridao Can you please take a look at this and respond when you are available?

**Comment by tridao** [comment]
Attention mask isn't supported (either in v1 or v2). I might implement it at some point but there are other priorities now.

**Comment by PeterL1n** [comment]
I thought mask is supported through `flash_attn_varlen_func`

https://github.com/Dao-AILab/flash-attention/blob/d30f2e1cd50185c98ed88c0684b4a603f15bee37/flash_attn/flash_attn_interface.py#L454C21-L454C21

**Comment by zhipeng93** [comment]
I have tested v1.0.7 and v2.0.4. The result turns out that none of them supports attention mask ---
- A: using flash attention with attention mask
- B: not using flash attention, with attention mask

The results of A and B are different.

**Comment by samvanstroud** [comment]
This paper might be relevant: https://arxiv.org/abs/2306.01160. 

There are several related issues:
- https://github.com/Dao-AILab/flash-attention/issues/530
- https://github.com/Dao-AILab/flash-attention/issues/506
- https://github.com/Dao-AILab/flash-attention/issues/424
- https://github.com/Dao-AILab/flash-attention/issues/342
- https://github.com/Dao-AILab/flash-attention/issues/307
- https://github.com/Dao-AILab/flash-attention/issues/242
- https://github.com/Dao-AILab/flash-attention/issues/127
- https://github.com/Dao-AILab/flash-attention/issues/119
- https://github.com/Dao-AILab/flash-attention/issues/17

I believe pytorch 2.1 will have a memory efficient attention implementation that supports arbitrary masks: https://github.com/pytorch/pytorch/issues/96099

**Comment by defei-coder** [comment]
@tridao Hello, I plan to add a bias mask in flashattention2. I noticed that in order to integrate the scale and add operations [scale_apply_exp2](https://github.com/Dao-AILab/flash-attention/blob/main/csrc/flash_attn/src/softmax.h#78) ,the scale is delayed until after the maximum value is calculated. I plan to support bias mask in the apply_mask_causal function, I think if a bias mask is supported, it seems that ffma optimization in scale_apply_exp2 can be cancelled. Using scale and bias can still benefit from FFMA, do you have any suggestions?

**Comment by zhangyipin** [comment]
[flash-attention](https://github.com/Dao-AILab/flash-attention/tree/main)/[flash_attn](https://github.com/Dao-AILab/flash-attention/tree/main/flash_attn)
/flash_attn_triton.py  you can use bias=-inf

**Comment by jaanli** [comment]
Anyone have tips on custom masks with flash attention for training? 

(I need this to train encoder-decoder models with variable-length sequences using non-causal masks.)

This came up in a recent article: https://www.yitay.net/blog/training-great-llms-entirely-from-ground-zero-in-the-wilderness

> The other striking thing is how little support these codebases have for large scale encoder-decoder training or even prefixLM training. To that end, even flash attention has consistently declined to provide support for prefixLM training (i.e., custom masks) despite reasonable demand on their github issues for whatever reason. 

Curious what this would take or if it is still out of scope for the flash attention library? 

Really grateful that this exists!! Just posting for visibility in case others have solved this problem :)

**Comment by tridao** [comment]
> Curious what this would take or if it is still out of scope for the flash attention library?

Not out of scope, it's just someone needs to go implement it :D

**Comment by jaanli** [comment]
Understood — thank you!! Will try using the varlen functions for now :)

**Comment by ardagoreci** [comment]
I was wondering if there has been any updates on this? AlphaFold3 uses a lot of attention pair biasing and it would be tremendously useful to computational biology if flash attention supported attention biasing!

**Comment by tridao** [comment]
> I was wondering if there has been any updates on this? AlphaFold3 uses a lot of attention pair biasing and it would be tremendously useful to computational biology if flash attention supported attention biasing!

Right, we still need someone to implement it.

**Comment by alexzhang13** [comment]
@tridao Was wondering, what needs to be done for this to be implemented (I'm assuming efficiently? otw it seems quite simple) 

I need a similar feature (arbitrary attention masks) but I figured I might take a stab at just implementing it if it still needs to be done.

**Comment by alexzhang13** [comment]
I've implemented a version of custom masking for FA2 in Triton: [https://github.com/alexzhang13/flashattention2-custom-mask](https://github.com/alexzhang13/flashattention2-custom-mask)

It suffices for my use case, but if something comes up where it's necessary to touch the FA3 code I may re-visit this.

**Comment by amyxlu** [comment]
> Attention mask isn't supported (either in v1 or v2). I might implement it at some point but there are other priorities now.

Seems like the FlashAttention class does [take in a `key_padding_mask` argument](https://github.com/Dao-AILab/flash-attention/blob/13403e81157ba37ca525890f2f0f2137edf75311/flash_attn/flash_attention.py#L27) in its forward method. What would be the difference between this and the attention mask to be implemented? Cc @tridao. Thanks!

**Comment by tridao** [comment]
As you can see in the code, key_padding_mask just removes elements from keys and values before passing to the flash attention kernel. There's no attention mask passed to the kernel.

**Comment by agshar96** [comment]
Hi Everyone,
Recently I published a paper in ENLSP Workshop@NEURips 2024, to address this problem, the paper can be found here: https://arxiv.org/pdf/2409.15097

I have the code, but its in a private repository currently, as I am still cleaning up the code. If someone wants to access this repo just send a mail to: agnivunoff96@gmail.com

Meanwhile, I realised that pytorch team already implemented a change which pretty much uses same method which I used. (I came up with my method independently for a university project the pytorch blog came around half a month **after** my university project).

Anyways, TL:DR - pytorch has now enabled custom masking of flash attention, you can find it here: https://pytorch.org/blog/flexattention/
(And, I am sad man, as my method will never be used)

**Comment by enszhou** [comment]
I aim to implement an attention mechanism that takes a continuous-valued mask (soft mask with values in [0, 1], not in the logarithmic domain as in attn_bias) and is fully differentiable with respect to the mask. Are there any existing implementations or references that support this functionality?

**Comment by vadimkantorov** [comment]
Are custom attention masks supported in any of FA's Triton/CuTE implementations? (and an especially important case is varlen impls, so is there any support for varlen attn_masks?)
